File: migration_agent/core/run_log.py
from __future__ import annotations

from datetime import datetime
from pathlib import Path


def create_run_log_path(output_path: Path) -> Path:
    output_path.mkdir(parents=True, exist_ok=True)
    timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    return output_path / f"migration-run-{timestamp}.log"


def append_log(log_path: Path | None, text: str) -> None:
    if log_path is None:
        return
    log_path.parent.mkdir(parents=True, exist_ok=True)
    with log_path.open("a", encoding="utf-8") as handle:
        handle.write(text)
        if not text.endswith("\n"):
            handle.write("\n")

File: src/Q3.MigrationAgent.Core/Logging/RunLog.cs
namespace Q3.MigrationAgent.Core.Logging;

public sealed class RunLog
{
    private readonly object _sync = new();
    public string CreateRunLogPath(string outputPath)
    {
        Directory.CreateDirectory(outputPath);
        var timestamp = DateTime.Now.ToString("yyyyMMdd-HHmmss");
        return Path.Combine(outputPath, $"migration-run-{timestamp}.log");
    }

    public void Append(string? logPath, string text)
    {
        if (string.IsNullOrWhiteSpace(logPath)) return;

        var content = text.EndsWith(Environment.NewLine, StringComparison.Ordinal)
            ? text
            : text + Environment.NewLine;

        lock (_sync)
        {
            Directory.CreateDirectory(Path.GetDirectoryName(logPath)!);

            using var stream = new FileStream(
                logPath,
                FileMode.Append,
                FileAccess.Write,
                FileShare.ReadWrite);

            using var writer = new StreamWriter(stream);
            writer.Write(content);
        }
    }
}